### PR TITLE
Fix module-scope constant declarations wrt type inference

### PR DESCRIPTION
**Pull request by kvark**
This reverts the part of #1679 which perhaps got in by accident?

Previously, we only agreed to infer types in local declarations, not module scopes. We also discussed type inference for global scope variables in #1550.

Also note that the current grammar doesn't allow module-scoped constants to omit types, anyway, so the examples are incorrect:
```
global_constant_decl
  : attribute_list* LET variable_ident_decl global_const_initializer?

variable_ident_decl
  : IDENT COLON attribute_list* type_decl
```

We can revisit this and discuss again, but we shouldn't have something in the spec that we didn't agree on.

### CODE REVIEW

**Comment by ben-clayton** [comment]
Argh. Tint has this implemented, with a stack of tests. I believe that there's considerable usage of this in downstream projects.

I agree this shouldn't have landed if the WG hadn't agreed, but yanking this out is going to cause a lot of pain.

Please can we discuss this?

**Comment by kvark** [comment]
I don't think a consideration for downstream projects is strong. WGSL had breaking changes recently, anyway, we aren't in MVP yet, this is nothing new.

But I'm happy to keep this open for discussion. One of the possible outcomes is that we close this PR and open a new one that fixes the grammar (in the opposite way).

I dug up the discussion in Rust on why they have global constants with types in https://github.com/rust-lang/rfcs/issues/1349, and it appears that inferring doesn't have obvious downsides if the type is fully inferred from the initializer (as opposed to anything else in the environment).

One small concern I could see is related to the future const expressions work. If constants are allowed to reference other constants, then it will no longer be obvious what the type of a declaration is:
```rust
let a = array<array<array<i32, 1>, 1, 1>(array<array<i32, 1>, 1>(array<i32, 1>(5)));
let b = a[0];
let c = b[0];
let d = c[0]; // what type is `d`?
```

So ideally we'd have a plan for this situation. For example, we could state that the type *has to be* provided (in the future) if the initializer mentions other constants.

**Comment by ben-clayton** [comment]
Thank you for being open to discussing this. Some of the downstream projects I'm referring to are Google internal, and will require fixing. It can be done, it's just not something I want us to have to do if we can instead agree to keep what's ended up in the spec.

> If constants are allowed to reference other constants, then it will no longer be obvious what the type of a declaration is

Sure, that's the cost of type inferencing. But omitting the type is optional, and if developers want to create complex `let` chains and can't be bothered to add a type, their confusion is their own doing.

**Comment by kainino0x** [comment]
I think I have argued in the past against type inference on module-scope `let`s, but mainly to reduce the scope of the feature.

> inferring doesn't have obvious downsides if the type is fully inferred from the initializer

Agree with this, and I agree with Ben's position. I think it's totally reasonable to have this kind of local inference at the module scope.

**Comment by kainino0x** [comment]
I guess this "Needs Discussion"?

File: wgsl/index.bs [hunk 1]
```diff
@@ -3027,8 +3028,8 @@ and the name denotes the value of that expression.
     // The golden ratio.
     let golden: f32 = 1.61803398875;
 
-    // The second unit vector for three dimensions, with inferred type.
-    let e2 = vec3<i32>(0,1,0);
+    // The second unit vector for three dimensions.
+    let e2: vec3<i32> = vec3<i32>(0,1,0);
```
Comment (dneto0) [inline wgsl/index.bs:3032]: Whoops.  Sorry about that.

File: wgsl/index.bs [hunk 2]
```diff
@@ -3018,6 +3018,7 @@ The name is available for use after the end of the declaration,
 until the end of the [SHORTNAME] program.
 
 A module-scope let-declared constant must be of [=constructible=] type.
+The type must always be specified.
```
Comment (dneto0) [inline wgsl/index.bs:3021]: This is good to have explicitly in the spec prose.

This is already enforced by the grammar rules, but that's probably too subtle.
(global_constant_decl uses a variable_ident_decl which has both the identifer and the type specifier.  The function-scope let-decl has a path with just the identifier.)

**Review by dneto0 — APPROVED** [review]

File: wgsl/index.bs (hunk 2 above)
Comment (dneto0) [reply wgsl/index.bs:3021]: (and I see you already mentioned that the grammar enforces it already.)

**Comment by dneto0** [comment]
My preference was (and still is) to allow type inference for module-scope constants.   I was fine with where the group landed before (to disallow it), and can live with either direction.

**Comment by jimblandy** [comment]
Let me rephrase what I was trying to say in that WGSL meeting long ago:

Some languages with type inference (ML, Haskell) examine _the entire program_ when sorting out the types. At first blush this sounds great - let's use all available information to help the programmer! But this means that one expression off in one function can influence the type of some other expression pages away, if there's data flow between them. In practice, when there is an error, it can be quite difficult to sort out why the compiler thinks a given expression must have a given type. The compiler can't tell which one of the constraints contributing to the conflict was the typo.

So a compromise used in Rust and other languages is to require function arguments and return types to be fully specified, and then infer types within each function's body separately, based solely on public declarations and that function's code. From the outside, a function's signature tells you everything you need to know; you never need to look inside it, as far as types are concerned. Limiting the range of type inference generally makes type errors less mysterious, by simplifying the influences.

In that context, since a global may be used from anywhere in the program, its type is analogous to the signature of a function: it's a public declaration. This is why Rust requires explicit types in global (`static` and `const`) declarations.

But in type-inferred languages, type inference usually applies to constructor expressions, too. WGSL doesn't work this way:

```
const DIAGONAL: (f32, f32) = (0.707, 0.707);        // Rust code
let diagonal: vec2<f32> = vec2<f32>(0.707, 0.707);  // WGSL
```

So this gives declarations a redundancy in WGSL that doesn't appear in Rust/ML/Haskell. So I can certainly see the argument for inferring globals' types, if every initializer is going to spell out the type anyway.

**Comment by dneto0** [comment]
The group agreed to allow type inference for module-scope let declarations during the [2021-09-07 meeting](https://docs.google.com/document/d/1NppRkM2UhtpQNwH-6V7Ek1jxzpzLUBEUyHk9Up63l5s/edit#heading=h.hp3f2zbslxr9).
